ngraph/core/reference/include/ngraph/runtime/reference/gather_nd.hpp

```diff
@@ -26,171 +28,117 @@ namespace ngraph
-                                                           params_axis_order);
-
-                size_t out_ndim = static_cast<size_t>(out_shape.size());
-                Coordinate out_start_corner(out_ndim, 0);
-                Coordinate out_end_corner(out_shape);
-                for (size_t i = 1; i < out_ndim; i++)
+
+                const auto first_slice_index_in_params = batch_dims + indices_shape.back();
+
+                if (!(first_slice_index_in_params <= params_shape.size()))
                 {
-                    out_end_corner[i] = 1;
+                    throw std::domain_error{
+                        "params_shape should have enough rank to be index by indices"};
                 }
-                Strides out_strides(out_ndim, 1);
-                AxisVector out_axis_order(out_ndim);
-                std::iota(out_axis_order.begin(), out_axis_order.end(), 0);
-                CoordinateTransform out_transform(
-                    out_shape, out_start_corner, out_end_corner, out_strides, out_axis_order);
-
-                Shape indices_shape_batch(indices_shape.begin() + batch_dims, indices_shape.end());
-                Shape params_shape_batch(params_shape.begin() + batch_dims, params_shape.end());
-                Shape output_shape_batch(out_shape.begin() + 1, out_shape.end());
-                auto out_coord_iter = out_transform.begin();
-                auto params_coord_iter = params_outer_transform.begin();
-                for (const Coordinate& indices_coord : indices_outer_transform)
+
+                const auto slice_shape = Shape(
+                    next(begin(params_shape), first_slice_index_in_params), end(params_shape));
+                const auto slice_size = shape_size(slice_shape);
+
+                const auto indices_offsets = [&] {
+                    std::vector<size_t> offsets{slice_size};
+                    const auto dims_begin = next(rbegin(params_shape), slice_shape.size());
+                    const auto dims_end = next(dims_begin, indices_shape.back() - 1);
+                    for (auto dim = dims_begin; dim != dims_end; ++dim)
+                    {
+                        const auto o = offsets.back() * *dim;
+                        offsets.push_back(o);
+                    }
+
+                    std::reverse(begin(offsets), end(offsets));
+                    return offsets;
+                }();
+
+                // algo sanity check - we need element offset in params for each axes
+                assert(indices_shape.back() == indices_offsets.size());
+
+                const auto batch_offset = indices_offsets.front() * params_shape[batch_dims];
+
+                const Shape k_1_indices(next(begin(indices_shape), batch_dims),
+                                        prev(end(indices_shape)));
+
+                const Shape k_1_params(next(begin(params_shape), batch_dims),
+                                       prev(end(params_shape)));
```

Comment:
Here we can avoid allocation extra vectors and work directly with indices_shape and params_shape. The k_1_indices is used only for shape size calculation. The k_1_params is later used inside the loop, and there we can read data directly from params_shape.

Reply:
Do we have something similar to span: https://en.cppreference.com/w/cpp/container/span?